### PR TITLE
Add debug logs for identifying slow jobs

### PR DESCRIPTION
**Pull request by rorybainfreetrade**
When testing lots of asynchronous code, `jobsComplete()` can become slow and it's often hard to find the source of the slowdown. This adds an opt in log that prints out a) the documents that database triggers ran on, b) the collections that had the most triggers, and c) the 20 slowest database triggers to run in milliseconds.

image

### CODE REVIEW

File: src/driver/RealtimeDatabase/InProcessRealtimeDatabase.ts [hunk 1]
```diff
@@ -356,7 +360,7 @@ export class InProcessRealtimeDatabaseRef
         eventType: "value",
     ): Promise<InProcessFirebaseRealtimeDatabaseSnapshot> {
         if (eventType !== "value") {
-            throw new Error('Only the "value" event type is supported.')
+            throw new Error("Only the \"value\" event type is supported.")
```
Comment (LukeInkster) [inline src/driver/RealtimeDatabase/InProcessRealtimeDatabase.ts:363]: What's the reason for this change? seems like it was better before

File: src/driver/RealtimeDatabase/InProcessRealtimeDatabase.ts [hunk 2]
```diff
@@ -537,10 +542,12 @@ export class InProcessRealtimeDatabase implements IFirebaseRealtimeDatabase {
...
         for (const observer of this.changeObservers) {
-            const job = new Promise((resolve) => {
-                setTimeout(async () => {
-                    resolve(observer.onChange({ before, after, data, delta }))
-                }, 1)
+            const start = performance.now()
+            const job = new Promise<IDatabaseChangePerformanceStats>((resolve) => {
+                observer.onChange({ before, after, data, delta }).then(() => resolve({
+                    path,
+                    durationMillis: Math.abs(performance.now() - start),
```
Comment (LukeInkster) [inline src/driver/RealtimeDatabase/InProcessRealtimeDatabase.ts:549]: `abs` again

**Review by LukeInkster — APPROVED** [review]

File: src/driver/PubSub/InProcessFirebasePubSub.ts [hunk 3]
```diff
@@ -84,23 +88,25 @@ export class InProcessFirebaseBuilderPubSub implements IFirebaseBuilderPubSub {
...
-                        handler(message, {
-                            timestamp: this.now().toISOString(),
-                        }),
-                    )
-                }, 1),
+                    handler(message, {
+                        timestamp: this.now().toISOString(),
+                    }).then(() => resolve({
+                        topicName: topicName,
+                        durationMillis: Math.abs(performance.now() - start),
```
[the remark this answers is not in the document]
Comment (rorybainfreetrade) [reply src/driver/PubSub/InProcessFirebasePubSub.ts:98]: giphy (3)

[the remark this answers is not in the document]
Comment (rorybainfreetrade) [reply src/driver/PubSub/InProcessFirebasePubSub.ts:98]: Ah yes, just need to switch the sum round

File: src/driver/AsyncJobs.ts [hunk 4]
```diff
@@ -1,39 +1,67 @@
...
+        const slowest = _.orderBy(stats, "durationMillis", ["desc"])
+        const pathsWithoutIds = stats.map(stat => {
+            const name = stat.path ?? stat.topicName ?? ""
+            const components = name.split("/")
+            return components.map((val, i) => i % 2 === 0 ? val : "*").join("/")
+        })
+        const uniqueCollections = new Set(pathsWithoutIds)
+        console.log(`jobsComplete stats: there were ${stats.length} paths updated`)
+        console.log(`jobsComplete stats: there were ${uniqueCollections.size} collections updated, the ones that were updated were:`)
+        console.log(`jobsComplete stats: the collections that were updated were: ${JSON.stringify(Array.from(uniqueCollections), null, 2)}`)
```
[the remark this answers is not in the document]
Comment (rorybainfreetrade) [reply src/driver/AsyncJobs.ts:62]: Good spot!

File: src/driver/RealtimeDatabase/InProcessRealtimeDatabase.ts (hunk 1 above)
Comment (rorybainfreetrade) [reply src/driver/RealtimeDatabase/InProcessRealtimeDatabase.ts:363]: Ah this is just automatic prettier/eslint, will revert

File: src/driver/PubSub/InProcessFirebasePubSub.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (LukeInkster) [reply src/driver/PubSub/InProcessFirebasePubSub.ts]: Do you? I'd have thought `start`, being the earlier time, would be the lower number. And to get a positive result you want to do higher - lower

[the remark this answers is not in the document]
Comment (rorybainfreetrade) [reply src/driver/PubSub/InProcessFirebasePubSub.ts]: Sorry yes, this is what we want: 
image
